Added Skybox Support

RuffledPlume opened the pull request:
Implemented support for Environments to specify a SkyboxID, when enabled skybox cubemap will be drawn before the scene does, replacing the fullscreen clear to the fogColor.

Added config option to toggle new skybox:
![image](https://github.com/user-attachments/assets/e05cc28f-9492-43cf-885a-d7f8907f0b86)

Skybox.json contains skybox mesh vertices and array of IDs that the enviorments.json can use to specify which skybox to be drawn when active, ID maps to subfolder: `textures\skybox\SKYBOX_ID`, texture face names map as follows:

| GL Target | Texture Name on disk|
| ------------- | ------------- |
| GL_TEXTURE_CUBE_MAP_POSITIVE_X | px |
| GL_TEXTURE_CUBE_MAP_NEGATIVE_X | nx |
| GL_TEXTURE_CUBE_MAP_POSITIVE_Y| py |
| GL_TEXTURE_CUBE_MAP_NEGATIVE_Y| ny |
| GL_TEXTURE_CUBE_MAP_POSITIVE_Z| pz |
| GL_TEXTURE_CUBE_MAP_NEGATIVE_Z| nz |


Transition Showcase:

https://github.com/user-attachments/assets/f563fb2c-26e7-466d-ba9d-cc220a3f4efa

General Showcase:

https://github.com/user-attachments/assets/3e6c2ed9-ab25-48f5-b50d-df8e9ec51772

RuffledPlume committed:
Skybox Support

    Added support for skyboxes within Environment
     - Environment Object can specify the optional property "skyboxId"
     - EnvironmentManager now blends different skyboxes together
       - Added getSkyboxProperties() to retrieve the values needed for skybox shader sampling

    environments.json
     - Overworld set to use "rs3_magicblue" to demonstrate skybox

    Added rs3_magicblue skybox
    Added rs3 greenlands skybox

    Note: In-order to support cubemap array min spec gl version would need to be #version 400

    TextureManager now checks & loads scene/skybox.json to parse the configuration/stored skybox textures
     - Creates a CubeMap Array & uploads all skybox faces into the array for use by scene & skybox shaders
     - config file contains array of ids which need to match up with the subdirectory name within textures/sandbox/*

    Added skybox_vert, skybox_frag & utils/skybox.glsl
     - new _vert/_frag shaders are used to draw the mesh stored in the sandbox.json file at the far clip of the camera
     - skybox.glsl provides the sampling of the skybox to allow for blending between different skyboxes

    Added Scene Chunk Boundary detection
    Helps improve weirdness when nearing the boundary of loaded scenes, were the ground near the camera fades out. Instead it'll remove any fog when nearing such a boundary making it clear we're about to load new chunks

Jin-Jiyunsun commented:
This still has textures with bad seams (my fault as I provided the edits originally)

[rs3_magicblue_seamfixes.zip](https://github.com/user-attachments/files/20739315/rs3_magicblue_seamfixes.zip)

Attached are the textures to fix this.

RuffledPlume committed:
Merge branch '117HD:master' into skybox

RuffledPlume committed:
Updated BlueSky Skybox with new Textures from Jin

RuffledPlume committed:
Merge remote-tracking branch '117HD/master' into skybox

RuffledPlume committed:
Merge remote-tracking branch 'origin/skybox' into skybox

RuffledPlume committed:
Merge branch 'master' into skybox

RuffledPlume committed:
Skybox Optimization

Moved Drawing of the Skybox to before the Memory Barrier, so that it's drawn whilst the Compute Task is running async utilising time that the GPU would have otherwise spent waiting.

RuffledPlume committed:
Merged Mark7625 Atlas support

RuffledPlume committed:
Merge branch '117HD:master' into skybox

Mark7625 committed:
Added Skybox Atlas

Mark7625 committed:
SkyBoxConfig

RuffledPlume committed:
Merge remote-tracking branch '117HD/master'

RuffledPlume committed:
Merge branch 'master' into skybox

RuffledPlume committed:
Check GL43 support before building skybox cubemap array

RuffledPlume committed:
Skybox Uniform Buffer WIP

RuffledPlume commented:
Refactored to go from a single vec4 uniform which use to be enough to hold: Blend, Offset, SkyboxIDx, NextSkyBoxIDx
To instead a new UniformBlock which is shared between the Skybox & Scene Shaders.
Takes two SkyboxConfig Structs which contain the Index & PostProcessing values.

Very much WIP, this is just the first pass to allow for expanded configuration making it into the SkyboxSample function

RuffledPlume committed:
Missing fix for skybox not drawing properly

Mark7625 committed:
Skybox Cleanup

- Reuse Texture Ids
- Moved Skybox loading to its own class

RuffledPlume committed:
Fixed bug in updateSkyboxUniformBuffer causing skybox to not draw when it should

RuffledPlume committed:
Merge branch 'master' into skybox

RuffledPlume committed:
Converted Skybox UBO to use UniformClass

RuffledPlume committed:
Removed Skybox Uniforms
Use Skybox & Global UBO instead of duplicated uniforms
Perform ViewProj Matrix multiply on CPU instead of in the vertex shader

Mark7625 committed:
Added more Texture Atlas Warnings

- Fixed the atlas margin for one of the images
- Cleaned up atlas code
- Added a dividable warning

Mark7625 committed:
Lowercase struct names

Mark7625 committed:
Added very basic rotation

Mark7625 committed:
Added skybox post for testing

aHooder committed:
Merge remote-tracking branch 'upstream/master' into pr558

aHooder committed:
Only use static imports for OpenGL 3.3 features

While slightly more annoying to use, this makes it easier to not
accidentally make use of newer OpenGL features in code which has to
support older versions (*cough* macOS).

aHooder committed:
Merge branch 'master' into pr558

aHooder committed:
Merge branch 'master' into pr558

Mark7625 committed:
Converted to use a quad

RuffledPlume committed:
Move Global UBO Property setting to be done first before any drawing commands are called

aHooder committed:
Move uniform till after scene program is active

aHooder committed:
Merge branch 'master' into pr558

RuffledPlume committed:
Fixed typo glUniformBlockBinding Global UBO for Skybox Shader Program

RuffledPlume commented:
> I'm guessing this PR will take a bit more time to finish, so I'll post a link to my alternative implementation here (https://github.com/aHooder/RLHD/tree/skybox) which uses equirectangular textures. Presumably the cubemap texture approach is faster, but that'd be nice to confirm :slightly_smiling_face: Also, I wonder, is there any reason for the cube-drawing approach rather than drawing a quad across the screen?
> 
> We're likely gonna want to bump up the size limit for the plugin a little bit so that we can include different skyboxes for different environments, but landing on a single solid one which works for the different seasonal options in the overworld would be good enough to merge this imo.

No reason at all, originally I was going to go with a full screen pass approach however had difficulty since I was still getting used to how the plugin was setup.

Also tbh, it was my first step back into graphics programming in general after a year and or so hiatus working on other core tech stuff. So not too attached to the approach in this.

Mark7625 committed:
Added Very basic roation

RuffledPlume committed:
Moved Scene X & Z AABB Uniform into Global UBO

RuffledPlume committed:
Merge branch 'skybox' of https://github.com/RuffledPlume/RLHD into skybox

aHooder committed:
Output sRGB from the sky fragment shader

Mark7625 committed:
Added Tint and roation and color fading over time

Mark7625 committed:
Added Skybox Editing Via Ui

Mark7625 committed:
Merge branch 'master' into skybox

aHooder committed:
Merge remote-tracking branch 'upstream/master' into pr558

aHooder committed:
Merge branch 'pr558-updates' into pr558

aHooder committed:
Merge remote-tracking branch 'upstream/master' into pr558

314159DD commented:
> I'm guessing this PR will take a bit more time to finish, so I'll post a link to my alternative implementation here (https://github.com/aHooder/RLHD/tree/skybox) which uses equirectangular textures. Presumably the cubemap texture approach is faster, but that'd be nice to confirm 🙂 Also, I wonder, is there any reason for the cube-drawing approach rather than drawing a quad across the screen?
> 
> We're likely gonna want to bump up the size limit for the plugin a little bit so that we can include different skyboxes for different environments, but landing on a single solid one which works for the different seasonal options in the overworld would be good enough to merge this imo.

Unfortunately i was not able to build this fork you shared with gradle 
`> Task :compileJava
C:\VST\RUNELITE\RLHD-skybox\src\main\java\rs117\hd\HdPlugin.java:2969: error: method does not override or implement a method from a supertype
        @Override
        ^
C:\VST\RUNELITE\RLHD-skybox\src\main\java\rs117\hd\HdPlugin.java:3149: error: incompatible types: possible lossy conversion from float to int
                                p.getPitchSin(),
                                             ^
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
2 errors
`

RuffledPlume commented:
> Unfortunately i was not able to build this fork you shared with gradle 
 The branch has gone abit stale, it needs master merging to fix the api issue

Mark7625 committed:
Merge branch 'master' into pr/558

Mark7625 commented:
> > I'm guessing this PR will take a bit more time to finish, so I'll post a link to my alternative implementation here (https://github.com/aHooder/RLHD/tree/skybox) which uses equirectangular textures. Presumably the cubemap texture approach is faster, but that'd be nice to confirm 🙂 Also, I wonder, is there any reason for the cube-drawing approach rather than drawing a quad across the screen?
> > We're likely gonna want to bump up the size limit for the plugin a little bit so that we can include different skyboxes for different environments, but landing on a single solid one which works for the different seasonal options in the overworld would be good enough to merge this imo.
> 
> Unfortunately i was not able to build this fork you shared with gradle `> Task :compileJava C:\VST\RUNELITE\RLHD-skybox\src\main\java\rs117\hd\HdPlugin.java:2969: error: method does not override or implement a method from a supertype @Override ^ C:\VST\RUNELITE\RLHD-skybox\src\main\java\rs117\hd\HdPlugin.java:3149: error: incompatible types: possible lossy conversion from float to int p.getPitchSin(), ^ Note: Some input files use or override a deprecated API. Note: Recompile with -Xlint:deprecation for details. Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output 2 errors `

I have updated the branch please turn off fog as it does not work correct with skybox right now and causes the wrong colour around the edges

Mark7625 committed:
scene_geom

Mark7625 committed:
Blend the fog with the sky colour

314159DD commented:
> You don’t need to merge anything from upstream manually — PR #558 is already fully up to date with the latest changes from the main repo.
> 
> If the skybox isn’t appearing, it’s likely because it’s currently restricted to certain environments only, so it won’t show up everywhere. Try heading to Varrock or Lumbridge, and you should see it working there.
> 
> The skybox setting is located under General, but it's enabled by default, so you shouldn’t need to change anything for it to work. If this does not work check the console see if any errors

I really appreciate the quick responses, thanks alot.
But im not able to get it to run apparently.

i did :
```
git clone https://github.com/117HD/RLHD.git
cd RLHD
git fetch origin pull/558/head:pr-558
git checkout pr-558
./gradlew clean jar
```
then moved this jar into a new folder i created _\.runelite\sideloaded-plugins_
set _--developer mode_ in the Runelite Client Configuration and launched, but i dont see a setting
Is this the correct approach even?

Mark7625 commented:
> > You don’t need to merge anything from upstream manually — PR #558 is already fully up to date with the latest changes from the main repo.
> > If the skybox isn’t appearing, it’s likely because it’s currently restricted to certain environments only, so it won’t show up everywhere. Try heading to Varrock or Lumbridge, and you should see it working there.
> > The skybox setting is located under General, but it's enabled by default, so you shouldn’t need to change anything for it to work. If this does not work check the console see if any errors
> 
> I really appreciate the quick responses, thanks alot. But im not able to get it to run apparently.
> 
> i did :
> 
> ```
> git clone https://github.com/117HD/RLHD.git
> cd RLHD
> git fetch origin pull/558/head:pr-558
> git checkout pr-558
> ./gradlew clean jar
> ```
> 
> then moved this jar into a new folder i created _.runelite\sideloaded-plugins_ set _--developer mode_ in the Runelite Client Configuration and launched, but i dont see a setting called Skybox and even in Lumbridge its not active Is this the correct approach even?

Have you tried uninstalling the plugin from the hub first ? sideloading may not override the hub one (i'm not 100% sure on that) also maybe try loading the plugin direct from `HdPluginTest ` this would be the best way to see it :)

314159DD commented:
> > > You don’t need to merge anything from upstream manually — PR #558 is already fully up to date with the latest changes from the main repo.
> > > If the skybox isn’t appearing, it’s likely because it’s currently restricted to certain environments only, so it won’t show up everywhere. Try heading to Varrock or Lumbridge, and you should see it working there.
> > > The skybox setting is located under General, but it's enabled by default, so you shouldn’t need to change anything for it to work. If this does not work check the console see if any errors
> > 
> > 
> > I really appreciate the quick responses, thanks alot. But im not able to get it to run apparently.
> > i did :
> > ```
> > git clone https://github.com/117HD/RLHD.git
> > cd RLHD
> > git fetch origin pull/558/head:pr-558
> > git checkout pr-558
> > ./gradlew clean jar
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then moved this jar into a new folder i created _.runelite\sideloaded-plugins_ set _--developer mode_ in the Runelite Client Configuration and launched, but i dont see a setting called Skybox and even in Lumbridge its not active Is this the correct approach even?
> 
> Have you tried uninstalling the plugin from the hub first ? sideloading may not override the hub one (i'm not 100% sure on that) also maybe try loading the plugin direct from `HdPluginTest ` this would be the best way to see it :)

i noticed i did not uninstall it in the launcher before placing my jar, no. So i went another round, but without luck.
I removed the 117 from the store, then restarted, placed my jar again in _.runelite\sideloaded-plugins_ and went again.
Now there is no 117 HD at all showing in the plugin configuration tab in runelite.

aHooder commented:
> set --developer mode in the Runelite Client Configuration and launched, but i dont see a setting called Skybox and even in Lumbridge its not active
RuneLite doesn't allow running the release client with `--developer-mode`. You'll have to either build RuneLite from source or run the plugin via the `HdPluginTest` class, which pulls in the latest snapshot build of RuneLite automatically.

aHooder commented:
> can't login

Ah, for that, see here https://github.com/runelite/runelite/wiki/Using-Jagex-Accounts

314159DD commented:
> > can't login
> 
> Ah, for that, see here https://github.com/runelite/runelite/wiki/Using-Jagex-Accounts

Thank you, im not really sure whats going on now but it doesnt work for me. I went back and forth but nothing.

I fixed the credential thing and also hooked my runelite profile with VM options. Also changed the project Structure to SDK 17 and Language Level 17
But when i start it now i see two 117 HD entries in the plugin list , none of them do anything though

So i went another round and now the module HdPluginTest is not specified.
I never worked with all of this so im at a loss here and my troubleshooting so far wasnt helping really and my gpt keeps misleading me so im loosing my patience slowly.
Class "rs117.hd.HdPluginTest" is not found in "hd" 
Deleted everything, went again.

Another round and im back to the 2 entries in the plugin list where neither do anything.

Can yall sanity check me here?
I do
git clone https://github.com/RuffledPlume/RLHD.git
cd there then checkout skybox
open the build.gradle file with intelli 
change my config so it loads my profile
and run?

i dont know whats going on at this point honestly

314159DD commented:
> If you see two 117 HD entries in the plugin list, that means either:
> 
>     * You're both sideloading the plugin and launching via `HdPluginTest`.
> 
>     * You're sideloading and have the plugin installed from the plugin hub.
>       I would uninstall it from the plugin hub, and delete the jar from `.runelite/sideloaded-plugins`, then run via `HdPluginTest` in IntelliJ, and it should just work without further manual configuration. `HdPluginTest` ensures you have the proper VM arguments passed as well, although it won't auto enable `--developer-mode` for you, if that's something you want. The plugin should run fine without that though if you run it via `HdPluginTest`.

Oh wow, appreciate your response time here.
And yes, it finally worked. Cant believe it was a simple thing like that...

Anyway i now got it and it works thats what counts - thanks alot mate!

Mark7625 committed:
Renamed water for skybox and added basic reflection

Mark7625 committed:
Revert "Renamed water for skybox and added basic reflection"

This reverts commit a4aa89c95c55592402dc0e37a63377ce1fd7bccf.

314159DD commented:
> If you see two 117 HD entries in the plugin list, that means either:
> 
>     * You're both sideloading the plugin and launching via `HdPluginTest`.
> 
>     * You're sideloading and have the plugin installed from the plugin hub.
>       I would uninstall it from the plugin hub, and delete the jar from `.runelite/sideloaded-plugins`, then run via `HdPluginTest` in IntelliJ, and it should just work without further manual configuration. `HdPluginTest` ensures you have the proper VM arguments passed as well, although it won't auto enable `--developer-mode` for you, if that's something you want. The plugin should run fine without that though if you run it via `HdPluginTest`.

Im wondering when will this feature hit the live version?
Tried to replicate the setup on a new installlation and it didnt work anymore unfortunately.

aHooder commented:
There's unfortunately too much to do in too little time. There's a lot of stuff I want to finish and release (including water, skybox, daylight cycle, tone mapping, bloom, fog, improved lights, improved shadows, just to name a few), but I don't want to push out something which isn't *quite* ready yet. We already get a fair bit of backlash even for changes which are well put together.

I completely understand that you'd like to see this in the plugin though, and thanks for expressing your interest in it :slightly_smiling_face: It might help us push for it a little bit earlier than it would otherwise land. For the next release, once 1.4.0 is stable, we need to focus on some new rendering changes in preparation for sailing (which should come with a big performance boost!). Unfortunately that means I can't really even give an ETA on when this PR might finally get merged. If someone is able to polish it up so it's in a state where it's overall better than what we currently have, then this might make it in sooner rather than later :slightly_smiling_face:

314159DD commented:
> There's unfortunately too much to do in too little time. There's a lot of stuff I want to finish and release (including water, skybox, daylight cycle, tone mapping, bloom, fog, improved lights, improved shadows, just to name a few), but I don't want to push out something which isn't _quite_ ready yet. We already get a fair bit of backlash even for changes which are well put together.
> 
> I completely understand that you'd like to see this in the plugin though, and thanks for expressing your interest in it 🙂 It might help us push for it a little bit earlier than it would otherwise land. For the next release, once 1.4.0 is stable, we need to focus on some new rendering changes in preparation for sailing (which should come with a big performance boost!). Unfortunately that means I can't really even give an ETA on when this PR might finally get merged. If someone is able to polish it up so it's in a state where it's overall better than what we currently have, then this might make it in sooner rather than later 🙂

I get it, great to hear you are still on it tho!
Me and my mates cant wait for it to drop. I showed it to them and the reception was great. They arent as technical and didnt want to do all the hassle necessary to have it ingame tho.

How realistic would a custom image as skybox be i wonder? Something along the lines of what [this user ](https://www.reddit.com/r/2007scape/comments/mzrgew/turns_out_all_you_need_is_a_skybox_to_make_osrs/)tried to do
